x-pack/auditbeat/module/system/socket/socket.go

```diff
@@ -0,0 +1,421 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux
+
+package socket
+
+import (
+	"fmt"
+	"net"
+	"os/user"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/OneOfOne/xxhash"
+	"github.com/gofrs/uuid"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/auditbeat/datastore"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	sock "github.com/elastic/beats/metricbeat/helper/socket"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/auditbeat/cache"
+	"github.com/elastic/gosigar/sys/linux"
+)
+
+const (
+	moduleName    = "system"
+	metricsetName = "socket"
+
+	bucketName              = "auditbeat.socket.v1"
+	bucketKeyStateTimestamp = "state_timestamp"
+
+	eventTypeState = "state"
+	eventTypeEvent = "event"
+
+	eventActionExistingSocket = "existing_socket"
+	eventActionSocketOpened   = "socket_opened"
+	eventActionSocketClosed   = "socket_closed"
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects data about sockets.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config Config
+	cache  *cache.Cache
+	log    *logp.Logger
+
+	netlink *sock.NetlinkSession
+	// TODO: Replace with process data collected in processes metricset
+	ptable    *sock.ProcTable
+	listeners *sock.ListenerTable
+
+	bucket    datastore.Bucket
+	lastState time.Time
+}
+
+// Socket represents information about a socket.
+type Socket struct {
+	Family       linux.AddressFamily
+	State        linux.TCPState
+	LocalIP      net.IP
+	LocalPort    int
+	RemoteIP     net.IP
+	RemotePort   int
+	Inode        uint32
+	Direction    sock.Direction
+	UID          uint32
+	Username     string
+	ProcessPID   int
+	ProcessName  string
+	ProcessError error
+}
+
+// newSocket creates a new socket out of a netlink diag message.
+func newSocket(diag *linux.InetDiagMsg) *Socket {
+	return &Socket{
+		Family:     linux.AddressFamily(diag.Family),
+		State:      linux.TCPState(diag.State),
+		LocalIP:    diag.SrcIP(),
+		LocalPort:  diag.SrcPort(),
+		RemoteIP:   diag.DstIP(),
+		RemotePort: diag.DstPort(),
```

Comment:
Question: Is `SrcIP` always local and `DstIP` remote? Also, should the Direction be set as well?

Reply:
I believe it should be. That's how its used in the Metricbeat `socket` metricset, `gosigar` calls it out [in a comment](https://github.com/elastic/gosigar/blob/master/sys/linux/inetdiag.go#L336), and I believe I even found the [line in the kernel code](https://elixir.bootlin.com/linux/v4.19.5/source/net/ipv4/tcp_ipv4.c#L1427) that sets it to be the local address even for incoming connections.

@andrewkroh can you confirm?

Reply:
On the direction: It's set [later in the code](https://github.com/elastic/beats/pull/8834/files/b32fc7a8ff935941783b08f2c7d0312430c1b6e3#diff-cf9498df2b7d9ebd4930c07961b0ecbdR373). The way the direction lookup works is the same [as in Metricbeat](https://github.com/elastic/beats/blob/master/metricbeat/module/system/socket/listeners.go#L108):

- Sockets with `RemotePort == 0` are `Listening` and added to a listener table
- Sockets that connect on a local port and protocol that has a listening socket in the listener table are `Incoming`
- Every other socket is `Outgoing`

Reply:
The only exception I know about has been reported in https://github.com/elastic/beats/issues/3306 which has not been fixed.